File: source/ch/cyberduck/ui/cocoa/UploadPromptModel.java
package ch.cyberduck.ui.cocoa;

/*
 *  Copyright (c) 2007 David Kocher. All rights reserved.
 *  http://cyberduck.ch/
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  Bug fixes, suggestions and comments should be sent to:
 *  dkocher@cyberduck.ch
 */

import ch.cyberduck.core.Path;
import ch.cyberduck.core.PathFilter;
import ch.cyberduck.core.Status;
import ch.cyberduck.core.Transfer;
import ch.cyberduck.ui.cocoa.foundation.NSAttributedString;
import ch.cyberduck.ui.cocoa.foundation.NSObject;

/**
 * @version $Id$
 */
public class UploadPromptModel extends TransferPromptModel {

    public UploadPromptModel(TransferPromptController c, Transfer transfer) {
        super(c, transfer);
    }

    /**
     * Filtering what files are displayed. Used to
     * decide which files to include in the prompt dialog
     */
    private PathFilter<Path> filter = new PromptFilter() {
        @Override
        public boolean accept(Path child) {
            log.debug("accept:" + child);
            if(child.exists()) {
                return super.accept(child);
            }
            return false;
        }
    };

    @Override
    protected PathFilter<Path> filter() {
        return filter;
    }

    @Override
    protected NSObject objectValueForItem(final Path item, final String identifier) {
        final NSObject cached = tableViewCache.get(item, identifier);
        if(null == cached) {
            if(identifier.equals(TransferPromptModel.WARNING_COLUMN)) {
                if(item.attributes.isFile()) {
                    if(item.getLocal().attributes.getSize() == 0) {
                        return tableViewCache.put(item, identifier, ALERT_ICON);
                    }
                    if(item.attributes.getSize() > item.getLocal().attributes.getSize()) {
                        return tableViewCache.put(item, identifier, ALERT_ICON);
                    }
                }
                return tableViewCache.put(item, identifier, null);
            }
            if(identifier.equals(TransferPromptModel.SIZE_COLUMN)) {
                return tableViewCache.put(item, identifier, NSAttributedString.attributedStringWithAttributes(Status.getSizeAsString(item.attributes.getSize()),
                        TableCellAttributes.browserFontRightAlignment()));
            }
        }
        return super.objectValueForItem(item, identifier);
    }
}

File: source/ch/cyberduck/ui/controller/UploadPromptModel.cs
// 
// Copyright (c) 2010-2013 Yves Langisch. All rights reserved.
// http://cyberduck.ch/
// 
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// Bug fixes, suggestions and comments should be sent to:
// yves@cyberduck.ch
// 

using ch.cyberduck.core;
using ch.cyberduck.core.transfer;
using org.apache.log4j;

namespace Ch.Cyberduck.Ui.Controller
{
    internal class UploadPromptModel : TransferPromptModel
    {
        private readonly Filter _filter;

        public UploadPromptModel(TransferPromptController controller, Transfer transfer) : base(controller, transfer)
        {
            _filter = new UploadPathFilter(transfer);
        }

        /// <summary>
        /// Filtering what files are displayed. Used to
        /// decide which files to include in the prompt dialog
        /// </summary>
        /// <returns></returns>
        public override Filter Filter()
        {
            return _filter;
        }

        public override object GetSize(Path path)
        {
            return path.attributes().getSize();
        }

        public override object GetWarningImage(Path path)
        {
            if (path.attributes().isFile())
            {
                if (path.attributes().getSize() == 0)
                {
                    return AlertIcon;
                }
                if (path.attributes().getSize() > path.getLocal().attributes().getSize())
                {
                    return AlertIcon;
                }
            }
            return null;
        }

        internal class UploadPathFilter : PromptFilter
        {
            protected static Logger Log = Logger.getLogger(typeof (UploadPathFilter).FullName);
            private readonly Transfer _transfer;

            public UploadPathFilter(Transfer transfer)
            {
                _transfer = transfer;
            }

            public override bool accept(object ap)
            {
                Path file = (Path) ap;
                if (_transfer.cache().lookup(file.getReference()) != null)
                {
                    return true;
                }
                return false;
            }
        }
    }
}